src/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs

```diff
@@ -345,5 +347,108 @@ private bool WaitForInputIdleCore(int milliseconds)
             }
             return idle;
         }
+
+        /// <summary>Checks whether the argument is a direct child of this process.</summary>
+        /// <remarks>
+        /// A child process is a process which has this process's id as its parent process id and which started after this process did.
+        /// </remarks>
+        private bool IsParentOf(Process possibleChild) =>
+            StartTime < possibleChild.StartTime
+            && Id == possibleChild.ParentProcessId;
+
+        /// <summary>
+        /// Get the process's parent process id.
+        /// </summary>
+        private unsafe int ParentProcessId
+        {
+            get
+            {
+                Interop.NtDll.PROCESS_BASIC_INFORMATION info = default;
+
+                using (SafeProcessHandle handle = GetProcessHandle(Interop.Advapi32.ProcessOptions.PROCESS_QUERY_INFORMATION))
+                {
+                    if (Interop.NtDll.NtQueryInformationProcess(handle, Interop.NtDll.PROCESSINFOCLASS.ProcessBasicInformation, &info, (uint)sizeof(Interop.NtDll.PROCESS_BASIC_INFORMATION), out _) != 0)
+                        throw new Win32Exception(SR.ProcessInformationUnavailable);
+
+                    return (int)info.InheritedFromUniqueProcessId;
+                }
+            }
+        }
+
+        private bool Equals(Process process) =>
+            Id == process.Id
+            && StartTime == process.StartTime;
+
+        private IEnumerable<Exception> KillTree()
+        {
+            // The process's structures will be preserved as long as a handle is held pointing to them, even if the process exits or 
+            // is terminated. A handle is held here to ensure a stable reference to the process during execution.
+            using (SafeProcessHandle handle = GetProcessHandle(Interop.Advapi32.ProcessOptions.PROCESS_QUERY_LIMITED_INFORMATION))
+            {
+                return KillTree(handle);
+            }
+        }
+
+        private IEnumerable<Exception> KillTree(SafeProcessHandle handle)
+        {
+            List<Exception> exceptions = new List<Exception>();
+
+            try
+            {
+                // Kill the process, so that no further children can be created.
+                //
+                // This method can return before stopping has completed. Down the road, could possibly wait for termination to complete before continuing.
+                Kill();
```

Comment:
Maybe we should do this already, calling WaitForExit should be enough.

Reply:
I do not know. To be certain it is probably necessary to wait on the process handle to be signaled (which means it's finished terminating) and pessimistically assume it could have spawned more processes. That seems a corner case we can perhaps ignore for now.

Reply:
Thanks, @tmds & @dnfadmin. Sounds like one of those many edge cases Dan mentioned that we'd discover.... :-)